File: test/browser/tests/code_highlighting.test.js
import { test } from "../test_helper.js"
import { expect } from "@playwright/test"
import { assertEditorContent } from "../helpers/assertions.js"

test.describe("Code highlighting", () => {
  test.beforeEach(async ({ page }) => {
    await page.goto("/")
    await page.waitForSelector("lexxy-editor[connected]")
    await page.waitForSelector("lexxy-toolbar[connected]")
  })

  test("ruby code is highlighted in editor", async ({ page, editor }) => {
    await editor.send("def hello_world")
    await editor.select("dev")
    await page.getByRole("button", { name: "Code" }).click()

    const languageSelect = page.locator("select[name=lexxy-code-language]")
    await expect(languageSelect).toHaveValue("plain")

    await languageSelect.selectOption("Ruby")

    await assertEditorContent(editor, async (content) => {
      await expect(content.locator("span.code-token__attr")).toContainText(
        "def",
      )
    })
  })
})

File: test/system/code_highlighting_test.rb
require "application_system_test_case"

class CodeHighlightingTest < ApplicationSystemTestCase
  setup do
    visit edit_post_path(posts(:empty))
  end

  test "ruby code is highlighted in editor and after saving" do
    find_editor.send "def hello_world"
    find_editor.select("dev")
    click_on "Code"
    assert_equal "plain", find("select[name=lexxy-code-language]").value

    select "Ruby", from: "lexxy-code-language"
    assert_selector "span.code-token__attr", text: "def"

    click_on "Update Post"

    # Verify the rendered output has syntax highlighting
    assert_selector "pre[data-language='ruby']"
    assert_selector "pre span.token.keyword", text: "def"
  end

  test "color highlights inside code blocks are preserved in rendered view" do
    find_editor.send "def hello_world"
    find_editor.select("dev")
    click_on "Code"
    select "Ruby", from: "lexxy-code-language"

    # Apply a background highlight to "hello"
    find_editor.select "hello"
    apply_highlight_option "background-color", 1

    click_on "Update Post"

    # Verify the rendered output preserves the highlight mark inside Prism tokens
    assert_selector "pre[data-language='ruby']"
    assert_selector "pre mark[style*='background-color']", text: "hello"
  end
end
